Adding OIDC provider (#1694)
sensu/sensu-docs: content/sensu-go/5.12/installation/auth.md

  - [Examples](#active-directory-configuration-examples)
  - [Specification](#active-directory-specification)
  - [Troubleshooting](#active-directory-troubleshooting)
- [OIDC](#oidc-authentication)
  - [OIDC configuration examples](#oidc-configuration-example)
  - [`oidc` attributes](#oidc-attributes)

>   - [Oktra](#okta)

Reviewer: ```suggestion
  - [Okta](#okta)
```

See the [LDAP troubleshooting](#ldap-troubleshooting) section.

## OIDC authentication

> The Sensu offers license-activated support for OIDC driver for using the OpenID Connect 1.0 protocol (OIDC) on top of the OAuth 2.0 protocol for RBAC authentication.

Reviewer: ```suggestion
The Sensu offers license-activated support for OIDC driver for using the OpenID Connect 1.0 protocol (OIDC) on top of the OAuth 2.0 protocol for RBAC authentication.

_NOTE: OIDC authentication is currently supported only via `sensuctl`. OIDC authentication for the Web UI will be added in a future release._
```

### OIDC configuration examples

{{< language-toggle >}}

{{< highlight yml >}}
---

> type: oidc

Reviewer: @ccressent is the type case sensitive here?

Reviewer 2: In the strictest sense, yes, because only `oidc` or `OIDC` will work (see [the typemap](https://github.com/sensu/sensu-enterprise-go/blob/master/api/authentication/v2/typemap.go#L31-L32)); a mixed case type will lead to the type not being recognized.

Author: leaving the example as lowercase.

api_version: authentication/v2
spec:
  additional_scopes:
  - groups
  - email
  client_id: a8e43af034e7f2608780
  client_secret: b63968394be6ed2edb61c93847ee792f31bf6216
  redirect_uri: http://127.0.0.1:8080/api/enterprise/authenication/v2/oidc/callback

>   server: https://localhost:9031

Reviewer: ```suggestion
  server: https://oidc.example.com:9031
```

  groups_claim: groups
  groups_prefix: 'oidc:'
  username_claim: email
  username_prefix: 'oidc:'
{{< /highlight >}}

{{< highlight json >}}
{
   "type": "oidc",
   "api_version": "authentication/v2",
   "spec": {
      "additional_scopes": [
         "groups",
         "email"
      ],
      "client_id": "a8e43af034e7f2608780",
      "client_secret": "b63968394be6ed2edb61c93847ee792f31bf6216",

>       " redirect_uri": "http://127.0.0.1:8080/api/enterprise/authenication/v2/oidc/callback",

Reviewer: ```suggestion
      "redirect_uri": "http://sensu-backend.example.com:8080/api/enterprise/authenication/v2/oidc/callback",
```

required     | true
type         | String
example      | {{< highlight shell >}}"client_id": "1c9ae3e6f3cc79c9f1786fcb22692d1f"{{< /highlight >}}

| client_secret  |      |
-------------|------
description  | The OIDC provider application "Client Secret" _NOTE: requires [registration of an application in the OIDC provider][27]._
required     | true
type         | String
example      | {{< highlight shell >}}"client_secret": "a0f2a3c1dcd5b1cac71bf0c03f2ff1bd"{{< /highlight >}}

| server |  |
-------------|------
description  | The location of the OIDC server you wish to authenticate against. _NOTE: Configuring with http will cause the connection to  be insecure._
required     | true
type         | String
example      | {{< highlight shell >}}"name_attribute": "https://sensu.oidc.provider.example.com"{{< /highlight >}}

| redirect_uri |   |
-------------|------
description  | Redirect URL to provide to the OIDC provider. Requires `/api/enterprise/authenication/v2/oidc/callback` _NOTE: only required for certain OIDC providers, such as Okta._
required     | false
type         | String

> example      | {{< highlight shell >}}"redirectURL": "http://localhost:8080/api/enterprise/authenication/v2/oidc/callback"{{< /highlight >}}

Reviewer: ```suggestion
example      | {{< highlight shell >}}"redirect_url": "http://sensu-backend.example.com:8080/api/enterprise/authenication/v2/oidc/callback"{{< /highlight >}}
```

Reviewer 2: It's `redirect_uri`

| groups_claim |   |
-------------|------
description  | The claim to use to form the associated RBAC groups. 
required     | false

> type         | Array

Reviewer 3: Testing this for sensu-puppet module using CI builds and seems `groups_claim` is a String.

Reviewer 2: The claim name is indeed going to be a string; the value held at that particular key needs to be an array of strings.

For example, if you've set up your external OIDC provider to put the groups as an array of strings in the 'my-groups' claim, it will ultimately send us some JSON in the form:

```
...
"my-groups": ["dev", "ops"],
...
```

and your Sensu OIDC config needs to include:
```
...
groups_claim: 'my-groups'
...
```

> type         | Array

Author: ```suggestion
type         | String
```

> example      | {{< highlight shell >}} "groups_claim": [ "ops", "dev" ] {{< /highlight >}}

Author: ```suggestion
example      | {{< highlight shell >}} "groups_claim": "ops" {{< /highlight >}}
```

| groups_prefix |   |
-------------|------
description  | A prefix to use to form the final RBAC groups if required.
required     | false
type         | String
example      | {{< highlight shell >}}"groups_prefix": "okta"{{< /highlight >}}

| username_claim |   |
-------------|------
description  | The claim to use to form the final RBAC user name.
required     | false
type         | String
example      | {{< highlight shell >}}"username_claim": "person"{{< /highlight >}}

| username_prefix |   |
-------------|------
description  | A prefix to use to form the final RBAC user name.
required     | false
type         | String
example      | {{< highlight shell >}}"username_prefix": "okta"{{< /highlight >}}

| additional_scopes |   |
-------------|------

> description  | Scopes to include in the claims, in addition to the default `email`, `openid` and `profile` scopes. _NOTE: only required for certain OIDC providers, such as Okta._

Reviewer: ```suggestion
description  | Scopes to include in the claims, in addition to the default `openid` scope. _NOTE: For most providers you'll want to include `groups`, `email` and `profile` in this list. Also, `groups` is not a standard claim and needs to be configured   for most external provider (see below)._
```

Reviewer 2: `username` is not a standard claim; maybe you meant `profile`?
You may want to note that `groups` is not a standard claim and needs to be created and configured on the external provider's side.

required     | false
type         | Array
example      | {{< highlight shell >}}"additional_scopes": ["groups", "email", "username"]{{< /highlight >}}

## Register an OIDC Application

To use OIDC for authentication requires registration of your Sensu Go 5.12.0 or later,
 as an "application". Please note the following instructions to
register an OIDC application for Sensu Enterprise based on your OIDC provider:

- [Okta](#okta)

> - [PingFederate](#pingfederate)

Reviewer: Remove the pingfederate link unless we're planning to cover it in an update to this PR

### Okta

#### Requirements

- Access to the Okta Administrator Dashboard
- Sensu Go 5.12.0 or later with a valid license

#### Create an Okta Application

1. From the Administrator Dashboard, select `Applications > Add Application > Create New App` to start the wizard.
2. Select the `Web` platform and `OpenID Connect` sign in method.
3. In General Settings enter an app name and (optionally) upload a logo.
4. In Configure OpenID Connect, add the following Redirect URI, without forgetting to replace DASHBOARD_URL with the URL to your dashboard: `{DASHBOARD_URL}/api/enterprise/authentication/v2/oidc/callback`
5. Click Save.
6. Head over to the Sign On page and click on the Edit button of the OpenID Connect ID Token section.
7. Enter the following information for the Groups claim attribute
  - First field: `groups`
  - Dropdown menu: `Regex`
  - Second field: `.*`
8. Click Save
9. Make sure to assign people and/or groups in the Assignments page

#### OIDC Driver Configuration

1. Add the `aadditional_scopes` configuration attribute in the [OIDC scope][26] and set the value to `[ "groups" ]`, just like this:

   >  `"additional_scopes": [ "groups" ]`

> 2. Add the `groups` to the `groups_claimed` array. For example an Okta group `dev` and we set a `groups_prefix` to `okta:`, we can setup RBAC objects to mention group `okta:dev` as needed

Author: ```suggestion
2. Add the `groups` to the `groups_claimed` string. For example an Okta group `dev` and we set a `groups_prefix` to `okta:`, we can setup RBAC objects to mention group `okta:dev` as needed
```

   >  `"additional_scopes": [ "groups" ]`

3. Add the `redirect_uri` configuration attribute in the [OIDC scope][26] and set
  the value to the Redirect URI configured at step 4 of
  [Create an Okta Application](#create-an-okta-application), just like this:
  
  > `"redirect_uri": "{DASHBOARD_URL}/api/enterprise/authentication/v2/oidc/callback"`

> 

Reviewer: TODO: add content re: using `sensuctl oidc login` to authenticate via OIDC.

      "server": "https://localhost:9031",
      "groups_claim": "groups",
      "groups_prefix": "oidc:",
      "username_claim": "email",
      "username_prefix": "oidc:"
   }
}
{{< /highlight >}}

{{< /language-toggle >}}

### RBAC for OIDC definition specification

#### `oidc` attributes

| client_id    |      |
-------------|------
description  | The OIDC provider application "Client ID" _NOTE: requires [registration of an application in the OIDC provider][27]._
required     | true
type         | String
example      | {{< highlight shell >}}"client_id": "1c9ae3e6f3cc79c9f1786fcb22692d1f"{{< /highlight >}}

| client_secret  |      |
-------------|------
description  | The OIDC provider application "Client Secret" _NOTE: requires [registration of an application in the OIDC provider][27]._
required     | true
type         | String
example      | {{< highlight shell >}}"client_secret": "a0f2a3c1dcd5b1cac71bf0c03f2ff1bd"{{< /highlight >}}

| server |  |
-------------|------
description  | The location of the OIDC server you wish to authenticate against. _NOTE: Configuring with http will cause the connection to  be insecure._
required     | true
type         | String

> example      | {{< highlight shell >}}"name_attribute": "https://sensu.oidc.provider.example.com"{{< /highlight >}}

Reviewer 2: s/name_attribute/server

5. Click Save.
6. Head over to the Sign On page and click on the Edit button of the OpenID Connect ID Token section.
7. Enter the following information for the Groups claim attribute
  - First field: `groups`
  - Dropdown menu: `Regex`
  - Second field: `.*`
8. Click Save
9. Make sure to assign people and/or groups in the Assignments page

#### OIDC Driver Configuration

1. Add the `aadditional_scopes` configuration attribute in the [OIDC scope][26] and set the value to `[ "groups" ]`, just like this:

   >  `"additional_scopes": [ "groups" ]`

2. Add the `groups` to the `groups_claimed` string. For example an Okta group `dev` and we set a `groups_prefix` to `okta:`, we can setup RBAC objects to mention group `okta:dev` as needed

   >  `"additional_scopes": [ "groups" ]`

3. Add the `redirect_uri` configuration attribute in the [OIDC scope][26] and set
  the value to the Redirect URI configured at step 4 of
  [Create an Okta Application](#create-an-okta-application), just like this:
  

>   > `"redirect_uri": "{DASHBOARD_URL}/api/enterprise/authentication/v2/oidc/callback"`

Reviewer 2: Needs to be `{BACKEND_URL}` instead of `{DASHBOARD_URL}`; e.g. for a system running locally: `localhost:8080/api/...` and not `localhost:3000/api/...`.

See the [LDAP troubleshooting](#ldap-troubleshooting) section.

## OIDC authentication

The Sensu offers license-activated support for OIDC driver for using the OpenID Connect 1.0 protocol (OIDC) on top of the OAuth 2.0 protocol for RBAC authentication.

> ### OIDC configuration examples

Reviewer 2: Need to update these examples with the `sensu-go` "format"; here is the one we were using to test during QA:

```
$ sensuctl auth info oidc --format yaml
type: OIDC
api_version: authentication/v2
metadata:
  name: oidc
spec:
  additional_scopes:
  - email
  - groups
  client_id: 0oa13ry4ypeDDBpxF357
  client_secret: DlArQRfND4BKBUyO0mE-TL2PWOVwyGjIO1fdk9gX
  groups_claim: groups
  groups_prefix: 'oidc:'
  redirect_uri: https://34.215.21.225:8080/api/enterprise/authentication/v2/oidc/callback
  server: https://sensu-cyril.okta.com
  username_claim: email
  username_prefix: 'oidc:'
```

Also, the struct describing the config is here if that helps:
https://github.com/sensu/sensu-enterprise-go/blob/master/api/authentication/v2/oidc.go#L19-L52

      datacenters: []
      subscriptions: []
      readonly: false
  metadata:
    name: oidc
{{< /highlight >}}

{{< highlight json >}}
{
  "type": "oidc",
  "api_version": "authentication/v2",
  "spec": {
    "oidc": {
      "additionalScopes": [ "groups" ],
      "clientId": "a8e43af034e7f2608780",
      "clientSecret": "b63968394be6ed2edb61c93847ee792f31bf6216",
      "insecure": false,
      "redirectURL": "http://127.0.0.1:4000/login/callback",
      "server": "https://localhost:9031",
      "roles": [
        {
          "name": "guests",
          "members": [
            "myorganization/guests"
          ],
          "datacenters": [
            "us-west-1"
          ],
          "subscriptions": [
            "webserver"
          ],
          "readonly": true
        },
        {
          "name": "operators",
          "members": [
            "myorganization/operators"
          ],
          "datacenters": [],
          "subscriptions": [],
          "readonly": false
        }
      ]
    },
  	"metadata": {
    	"name": "oidc"
  	}
	}
}
{{< /highlight >}}

{{< /language-toggle >}}

### RBAC for OIDC definition specification

#### `oidc` attributes

> additionalScopes | 

Reviewer 2: `additional_scopes` now

-----------------|------

> description      | Scopes to include in the claims, in addition to the default `email`, `openid` and `profile` scopes. _NOTE: only required for certain OIDC providers, such as Okta._

Reviewer 2: There is now only the `openid` scope that's there by default (in line with the OIDC spec). If the user needs `email`, `profile`, ... or anything else they need to add it explicitly.

required         | false
type             | Array
example          | {{< highlight shell >}}"additionalScopes": [ "groups" ]{{< /highlight >}}

> clientId     | 

Reviewer 2: `client_id` now.

-------------|------
description  | The OIDC provider application "Client ID" _NOTE: requires [registration of an application in the OIDC provider][4]._
required     | true
type         | String
example      | {{< highlight shell >}}"clientId": "a8e43af034e7f2608780"{{< /highlight >}}

> clientSecret | 

Reviewer 2: `client_secret` now.

-------------|------
description  | The OIDC provider application "Client Secret" _NOTE: requires [registration of an application in the OIDC provider][4]._
required     | true
type         | String
example      | {{< highlight shell >}}"clientSecret": "b63968394be6ed2edb61c93847ee792f31bf6216"{{< /highlight >}}

> insecure     | 

Reviewer 2: Not here (yet?) for `sensu-go`

Reviewer: @ccressent seems like `client_secret` is now part of the spec. Will you please confirm we should leave this in and ensure `cilentSecret` is changed to `client_secret` ?

Reviewer 2: `client_secret` is indeed there and mandatory.
My comment above was about the `insecure` field not being supported.

-------------|------
description  | Determines whether or not to skip SSL certificate verification (e.g. for self-signed certificates).
required     | false
type         | Boolean
default      | false
example      | {{< highlight shell >}}"insecure": true{{< /highlight >}}

> redirectURL  | 

Reviewer 2: Called `redirect_uri` now.

-------------|------
description  | Redirect URL to provide to the OIDC provider. _NOTE: only required for certain OIDC providers, such as Okta._
required     | false
type         | String
example      | {{< highlight shell >}}"redirectURL": "http://localhost:4000/login/callback"{{< /highlight >}}

server       | 
-------------|------
description  | The location of the OIDC server you wish to authenticate against.
required     | true
type         | String
example      | {{< highlight shell >}}"server": "https://localhost:9031"{{< /highlight >}}

> roles        | 

Reviewer 2: Removed in `sensu-go`.

-------------|------
description  | An array of`roles` definitions.
required     | true
type         | Array
example      | {{< highlight shell >}}"roles": [
  {
    "name": "guests",
    "members": [
      "myorganization/guests"
    ],
    "datacenters": [
      "us-west-1"
    ],
    "subscriptions": [
      "webserver"
    ],
    "readonly": true
  },
  {
    "name": "operators",
    "members": [
      "myorganization/operators"
    ],
    "datacenters": [],
    "subscriptions": [],
    "readonly": false
  }
]
{{< /highlight >}}

#### `roles` attributes

Please see the [RBAC definition specification][25] for information on how to
configure RBAC roles.

## Register an OIDC Application

To use OIDC for authentication requires registration of your Sensu Enterprise

> Dashboard as an "application". Please note the following instructions to

Reviewer 2: It's not only for the dashboard now; in fact for 5.12 we only support CLI login. The interface in the WebUI is coming later.

register an OIDC application for Sensu Enterprise based on your OIDC provider:

- [Okta](#okta)
- [PingFederate](#pingfederate)

### Okta

#### Requirements

- Access to the Okta Administrator Dashboard

> - Sensu Enterprise Dashboard 2.9.0 or later

Reviewer: ```suggestion
- Sensu Go 5.12.0 or later
```

> - Sensu Enterprise Dashboard 2.9.0 or later

Reviewer 2: That's probably "Sensu Go 5.12 or later, with a valid license" now, or something in that vein.

#### Create an Okta Application

1. From the Administrator Dashboard, select `Applications > Add Application > Create New App` to start the wizard.
2. Select the `Web` platform and `OpenID Connect` sign in method.
3. In General Settings enter an app name and (optionally) upload a logo.

> 4. In Configure OpenID Connect, add the following Redirect URI, without forgetting to replace DASHBOARD_URL with the URL to your dashboard: `{DASHBOARD_URL}/login/callback`

Reviewer 2: The URL would now be something like `{BACKEND_URL}/api/enterprise/authentication/v2/oidc/callback`, eg: `https://my-sensu-api.example.com:8080/api/enterprise/authentication/v2/oidc/callback`.

5. Click Save.
6. Head over to the Sign On page and click on the Edit button of the OpenID Connect ID Token section.
7. Enter the following information for the Groups claim attribute
  - First field: `groups`
  - Dropdown menu: `Regex`
  - Second field: `.*`
8. Click Save
9. Make sure to assign people and/or groups in the Assignments page

#### OIDC Driver Configuration

> 1. Add the `additionalScopes` configuration attribute in the [OIDC scope][26] and set the value to `[ "groups" ]`, just like this:

Reviewer 2: Instead now we would set `groupsClaim` to `groups` and `groupsPrefix` to whatever they want the associated group to show up as in RBAC.

For example, if my user on Okta is in group `dev` and we set the `groupsPrefix` to `okta:`, we'd setup the RBAC objects to mention group `okta:dev` as needed.

   >  `"additionalScopes": [ "groups" ]`
2. Add the `redirectURL` configuration attribute in the [OIDC scope][26] and set
  the value to the Redirect URI configured at step 4 of
  [Create an Okta Application](#create-an-okta-application), just like this:
  
  > `"redirectURL": "{DASHBOARD_URL}/login/callback"`

> ### PingFederate

Reviewer 2: I haven't personally tried PingFederate, but @palourde did!

type: oidc
api_version: authentication/v2
spec:
  additional_scopes:
  - groups
  - email
  clientId: a8e43af034e7f2608780
  clientSecret: b63968394be6ed2edb61c93847ee792f31bf6216

>   redirectURL: http://127.0.0.1:8080/api/enterprise/authenication/v2/oidc/callback

Reviewer 2: s/authenication/authentication

(also elsewhere in the file)

>   redirectURL: http://127.0.0.1:8080/api/enterprise/authenication/v2/oidc/callback

Reviewer 2: `redirect_uri`

description  | The OIDC provider application "Client ID" _NOTE: requires [registration of an application in the OIDC provider][27]._
required     | true
type         | String
example      | {{< highlight shell >}}"client_id": "1c9ae3e6f3cc79c9f1786fcb22692d1f"{{< /highlight >}}

| client_secret  |      |
-------------|------
description  | The OIDC provider application "Client Secret" _NOTE: requires [registration of an application in the OIDC provider][27]._
required     | true
type         | String
example      | {{< highlight shell >}}"client_secret": "a0f2a3c1dcd5b1cac71bf0c03f2ff1bd"{{< /highlight >}}

| server |  |
-------------|------
description  | The location of the OIDC server you wish to authenticate against. _NOTE: Configuring with http will cause the connection to  be insecure._
required     | true
type         | String
example      | {{< highlight shell >}}"name_attribute": "https://sensu.oidc.provider.example.com"{{< /highlight >}}

| redirect_uri |   |
-------------|------
description  | Redirect URL to provide to the OIDC provider. Requires `/api/enterprise/authenication/v2/oidc/callback` _NOTE: only required for certain OIDC providers, such as Okta._

> required     | false

Reviewer 2: `redirect_uri` is required

type         | String
example      | {{< highlight shell >}}"redirectURL": "http://localhost:8080/api/enterprise/authenication/v2/oidc/callback"{{< /highlight >}}

| groups_claim |   |
-------------|------

> description  | The claim to use to form the associated RBAC groups. 

Reviewer 2: You may want to add that the value held by that claim _must_ to be an array of strings

[22]: #user-search-attributes
[23]: #active-directory-metadata-attributes
[24]: #metadata-attributes

> [25]: #

Reviewer 4: Was this going to be used for something? Doesn't appear to be used on the page.
